### DISCUSSION TITLE
How to make Sidekiq faster

### DISCUSSION THREAD

**Discussion by todoneunl** [mastodon/mastodon#19797] [Q&A]
Sidekiq on my servers have problems coping with the strong increase of new Mastodon users in the fediverse. Especially the default queue has a latency in the afternoon and evening (EU time) between 15 and 30 minutes. Let me describe the way I configured Sidekiq:

I have thee processes created for systemd. A high priority one for the default queue, a low priority one for the pull queue and a normal priority one for the other queues.

High: 
```
Environment="DB_POOL=60"
Environment="MALLOC_ARENA_MAX=3"
ExecStart=/home/mastodon/.rbenv/shims/bundle exec sidekiq -c 60 -q default,8
```

Normal:
```
Environment="DB_POOL=25"
Environment="MALLOC_ARENA_MAX=2"
ExecStart=/home/mastodon/.rbenv/shims/bundle exec sidekiq -c 25 -q push,4 -q mailers,2 -q scheduler
```

Low: 
```
Environment="DB_POOL=15"
Environment="MALLOC_ARENA_MAX=2"
ExecStart=/home/mastodon/.rbenv/shims/bundle exec sidekiq -c 15 -q pull
```

This all worked great until more and more people joined the network. I've closed registrations, but that doesn't seem to do anything. Pgbouncer is configured, increased ulimit increased and the [TCP Port Range](https://www.cyberciti.biz/tips/linux-increase-outgoing-network-sockets-range.html). Still no affect. The site itself (Puma) works fine. I increased the default queue to 100 or 200, but doesn't change anything. I even tried 1000, but that made the Sidekiq process crash.

Hopefully someone can help me of give more tips? There must be somewhere a bottleneck. The server I used as example is todon.nl. Todon.eu is on smaller VPS and the Sidekiq config on there is a bit lower (45 for default queue). I can upgrade todon.eu to the same VPS type as todon.nl, but I'm wondering if this solves the problem, because todon.nl has also issues.

Running up-to-date Debian 11 (bullseye) on a 10 core, 60GB RAM VPS with SSD (todon.nl) and 8 core, 30GB RAM with SSD (todon.eu). Running Mastodon 3.5.3.

**Comment by todoneunl** [comment]
In the meantime it's getting worse and worse. Help!
![afbeelding](https://user-images.githubusercontent.com/41594439/200176881-f366f9c1-f6e0-40b9-8341-aa3734acdcd3.png)

**Reply by todoneunl** [reply]
I fixed it. With help from @ mdbraber @ mastodon.nl. 

This is what I did.

I already had those three Sidekiq processes and the one with the default queue was the problem as you can see. Pull queue is also high sometimes, but that one is not so important (and normally processed very quickly) and won't really effect the user experience. But you can also do this for other queues of course, when you have the CPU and RAM. So I multiplied the high priority one for the default queue (that I named /etc/systemd/system/mastodon-sidekiq-high.service). Modify below how to for your own situation.

As root:

```
~# cp -a /etc/systemd/system/mastodon-sidekiq-high.service /etc/systemd/system/mastodon-sidekiq-high-@.service
~# vi /etc/systemd/system/mastodon-sidekiq-high-@.service
```

Change it with your preferred editor to this:

```
[Unit]
Description=mastodon-sidekiq-high-%i
After=network.target

[Service]
Type=simple
User=mastodon
WorkingDirectory=/home/mastodon/live
Environment="RAILS_ENV=production"
Environment="DB_POOL=30"
Environment="MALLOC_ARENA_MAX=3"
ExecStart=/home/mastodon/.rbenv/shims/bundle exec sidekiq -c 30 -q default,8
TimeoutSec=15
Restart=always

[Install]
WantedBy=multi-user.target
```

Save and quit. Let's continue:

```
~# systemctl stop mastodon-sidekiq-high.service
~# systemctl disable mastodon-sidekiq-high.service
~# systemctl start mastodon-sidekiq-high-@1.service
~# systemctl start mastodon-sidekiq-high-@2.service
~# systemctl start mastodon-sidekiq-high-@3.service
```

You can start as much as you want. Take look to your load average and RAM with htop what your VPS/server can handle. To make sure there are also a minimum started after reboot, do this:

```
~# systemctl enable mastodon-sidekiq-high-@1.service
~# systemctl enable mastodon-sidekiq-high-@2.service
```

This is it! Advise is to stop and start those mastodon-sidekiq-high-@ services/processes on demand, because it can effect performance of other tasks on your VPS/server. If would be even better if someone could write a script to start and stop those systemd services automatically.

**Reply by offbyone** [reply]
So the key was to split out the work between >1 process, rather than threads?

**Reply by todoneunl** [reply]
Yes.

> So the key was to split out the work between >1 process, rather than threads?

**Reply by maop** [reply]
Thanks a lot ! I was having latency of 3h ! Didn't know you could have multiple instances of sidekiq. Nice trick.

**Reply by ChadOhman** [reply]
Thank you @joenepraat. This solved my instance's three hour latency!

**Comment by offbyone** [comment]
I'm in the same boat, and nothing I've tried has worked for me either; we upgraded our instance to 3.5.3 from 3.4.6 yesterday, and it took almost 20 hours to burn through the backlog, but by this morning we're back up again.

Sidekiq utilization is way lower than I expect, given that there are waiting tasks: 
<img width="1185" alt="image" src="https://user-images.githubusercontent.com/181693/200182336-6cd6aef7-19be-4cec-a083-727b1901f31b.png">

My queue latencies are climbing continuously:

<img width="1194" alt="image" src="https://user-images.githubusercontent.com/181693/200182373-431ecd5e-c834-4d6c-bfc4-362624c908c6.png">

I've got the Sidekiq process for the default queue configured with 55 pooled connections to the DB, and a concurrency of 50. The database was tuned using pgTune to allow for 200 connections, and since that time I haven't been getting the `[7495] FATAL:  sorry, too many clients already` error that dogged my migration at the start.

**Comment by RaitoBezarius** [comment]
Maybe it would be timely to collect `perf` (https://perf.wiki.kernel.org/index.php/Main_Page) data on your VPSes so performance experts can chime in at some point and see what can be done.

I would advise running something like https://www.netdata.cloud/ to get performance monitoring on your VPS with enough metrics, configured for PostgreSQL. I do not think they have Sidekiq module, though, it can be added fairly easy I think.

(I even think that it would be a great showcase of Netdata, so maybe Netdata people may be interested working on that case.)

**Reply by offbyone** [reply]
I'm open to suggestions about how I would get that information meaningfully; my site is running in docker using compose, with similar configuration to what the OP mentions, but obviously subtle differences. I'm not a Linux perf expert, so I am not sure how to _gather_ that info, much less how to interpret it.

**Reply by RaitoBezarius** [reply]
Alas, I do not think that Linux containers ecosystem is sufficiently mature to reuse `perf` tooling without forking to eBPF or advanced observability tooling, the closest thing I know about is https://gendignoux.com/blog/2019/11/09/profiling-rust-docker-perf.html which would need to be adapted to profile the Ruby process of Mastodon. An interesting thing about having the process in the same namespace is you can profile them altogether and collect coherent information (i.e. PostgreSQL/Sidekiq, etc.)

Of course, a Linux container expert may prove me wrong.

Anyway, wrt to the data collected, I think posting it here is enough! As we will all able to work together to interpret it and provide for leads on how to optimize performance.

**Comment by nmalcolm** [comment]
Increase DB_POOL way above the thread count.

```
Environment="DB_POOL=200"
ExecStart=/home/mastodon/.rbenv/shims/bundle exec sidekiq -c 5 -q default
```

**Reply by offbyone** [reply]
How does that work?

**Reply by nmalcolm** [reply]
> How does that work?

Extremely well.

![image](https://user-images.githubusercontent.com/973100/200638789-c4cd12e5-a6ef-4236-8ede-6b4592cd7e84.png)

**Reply by offbyone** [reply]
No, I mean "what is the mechanism by which that improves performance?"

**Reply by nmalcolm** [reply]
> No, I mean "what is the mechanism by which that improves performance?"

As far as I'm aware, there's an internal pool of database connections and if it's set too low then sidekiq just waits until one becomes free. I've been dealing with this issue for a few days and the backlog reached 650,000 on the default queue causing a 10 hour lag. Now we're blazing though the backlog.

**Reply by todoneunl** [reply]
According the [Mastodon documentation](https://docs.joinmastodon.org/admin/scaling/#concurrency): "Please mind that each threads needs to be able to connect to the database, which means that the database pool needs to be large enough to support all the threads. The database pool size is controlled with the DB_POOL environment variable and must be at least the same as the number of threads."

I understand that every thread needs at least one database connection. So setting to the same number is fine I think, although it can not hurt to make the `DB_POOL` higher. But than the number of `max_connections` in postgresql.conf needs at least to have the same number. I always learned that 200 is quite high for Postgresql. Or you need to [use Pgbouncer](https://docs.joinmastodon.org/admin/scaling/#pgbouncer) and than you need to increase the `max_client_conn` in pgbouncer.ini.
